File: М11_ClusterVMS/clustervms-go/cluster/placement.go
package cluster

// Lesson 5 — placement onto Nodes: capacity measured, constraints as
// labels, the placement STORED in Variables (placement/<camera>), and one
// rule with a property test: adding a Node moves nothing.
//
// Rebalance is explicit, budgeted, observable, interruptible, with a dead
// band — and each move is the one two-writer operation in the module, so
// the caller performs it with the epoch: stop on the old Node, start on
// the new, never both at once without a fence.

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// Labels is a set: "vlan:cctv-b". A camera's labels must be a subset of
// its Node's.
type Labels map[string]bool

func NewLabels(ls ...string) Labels {
	out := Labels{}
	for _, l := range ls {
		out[l] = true
	}
	return out
}

func (a Labels) SubsetOf(b Labels) bool {
	for l := range a {
		if !b[l] {
			return false
		}
	}
	return true
}

type PCamera struct {
	ID     int64
	Load   float64 // in units of the probe's per-pipeline increment I
	Labels Labels
}

type PNode struct {
	ID       string
	Capacity float64 // (budget − B) / I from М10's shard-memory-probe
	Labels   Labels
}

type Placement struct {
	Node   string
	Reason string
	Rev    int64
}

type Move struct {
	Camera   int64
	From, To string
}

// Placer keeps placements in Variables under placement/<camera_id>: small,
// one writer (the placement service), exact lookups only.
type Placer struct {
	Nodes  map[string]PNode
	Vars   Variables
	Placed map[int64]Placement
}

func NewPlacer(nodes map[string]PNode, v Variables) (*Placer, error) {
	p := &Placer{Nodes: nodes, Vars: v, Placed: map[int64]Placement{}}
	paths, err := v.List("placement/")
	if err != nil {
		return nil, err
	}
	for _, path := range paths {
		items, _, err := v.Get(path)
		if err != nil {
			return nil, err
		}
		if items == nil || items["node"] == "" {
			continue
		}
		id, _ := strconv.ParseInt(strings.SplitN(path, "/", 2)[1], 10, 64)
		p.Placed[id] = Placement{items["node"], items["reason"], atoi(items["rev"])}
	}
	return p, nil
}

func (p *Placer) store(cam int64, pl Placement) error {
	path := fmt.Sprintf("placement/%d", cam)
	_, idx, err := p.Vars.Get(path)
	if err != nil {
		return err
	}
	if _, err := p.Vars.Put(path, Items{"node": pl.Node, "reason": pl.Reason, "rev": itoa(pl.Rev)}, idx); err != nil {
		return err
	}
	p.Placed[cam] = pl
	return nil
}

func (p *Placer) rev() (int64, error) {
	items, _, err := p.Vars.Get("placement")
	if err != nil {
		return 0, err
	}
	rev := atoi(items["rev"]) + 1
	_, err = p.Vars.Put("placement", Items{"rev": itoa(rev)}, NoCAS)
	return rev, err
}

func (p *Placer) nodeIDs() []string {
	ids := make([]string, 0, len(p.Nodes))
	for id := range p.Nodes {
		ids = append(ids, id)
	}
	sort.Strings(ids) // deterministic where Python relied on insertion order
	return ids
}

func (p *Placer) LoadOf(node string, cameras map[int64]PCamera) float64 {
	var sum float64
	for c, pl := range p.Placed {
		if cam, ok := cameras[c]; ok && pl.Node == node {
			sum += cam.Load
		}
	}
	return sum
}

func (p *Placer) Eligible(cam PCamera) []PNode {
	var out []PNode
	for _, id := range p.nodeIDs() {
		if cam.Labels.SubsetOf(p.Nodes[id].Labels) {
			out = append(out, p.Nodes[id])
		}
	}
	return out
}

// Place places ONE new camera; existing placements are never touched.
// nil means "the system is full" — never "Node 3 is full".
func (p *Placer) Place(cam PCamera, cameras map[int64]PCamera) (*Placement, error) {
	if pl, ok := p.Placed[cam.ID]; ok {
		return &pl, nil
	}
	var best *PNode
	bestFree := 0.0
	eligible := p.Eligible(cam)
	for i := range eligible {
		n := eligible[i]
		free := n.Capacity - p.LoadOf(n.ID, cameras)
		if free >= cam.Load && free > bestFree {
			best, bestFree = &n, free
		}
	}
	if best == nil {
		return nil, nil
	}
	rev, err := p.rev()
	if err != nil {
		return nil, err
	}
	pl := Placement{best.ID, fmt.Sprintf("most free capacity (%.1f) among %d eligible", bestFree, len(eligible)), rev}
	return &pl, p.store(cam.ID, pl)
}

func (p *Placer) Remove(cam int64) error {
	if _, ok := p.Placed[cam]; !ok {
		return nil
	}
	rev, err := p.rev()
	if err != nil {
		return err
	}
	path := fmt.Sprintf("placement/%d", cam)
	_, idx, err := p.Vars.Get(path)
	if err != nil {
		return err
	}
	if _, err := p.Vars.Put(path, Items{"node": "", "reason": "removed", "rev": itoa(rev)}, idx); err != nil {
		return err
	}
	delete(p.Placed, cam)
	return nil
}

// AddNode: nothing moves. That is the rule.
func (p *Placer) AddNode(n PNode) { p.Nodes[n.ID] = n }

// RetireNode is a Node retired ON PURPOSE (a failover moves the Node, not
// its cameras). Returns the cameras that found no home.
func (p *Placer) RetireNode(node string, cameras map[int64]PCamera) ([]int64, error) {
	delete(p.Nodes, node)
	var orphans []int64
	for c, pl := range p.Placed {
		if pl.Node == node {
			orphans = append(orphans, c)
		}
	}
	sort.Slice(orphans, func(i, j int) bool { return orphans[i] < orphans[j] })
	var homeless []int64
	for _, c := range orphans {
		delete(p.Placed, c)
		pl, err := p.Place(cameras[c], cameras)
		if err != nil {
			return nil, err
		}
		if pl == nil {
			homeless = append(homeless, c)
		}
	}
	return homeless, nil
}

// Rebalance moves at most budget cameras from the fullest Node to the
// emptiest, stopping inside the dead band. Every move carries a reason and
// a revision.
func (p *Placer) Rebalance(cameras map[int64]PCamera, budget int, deadBand float64) ([]Move, error) {
	var moves []Move
	for i := 0; i < budget; i++ {
		ids := p.nodeIDs()
		if len(ids) == 0 {
			break
		}
		loads := map[string]float64{}
		hi, lo := ids[0], ids[0]
		for _, id := range ids {
			loads[id] = p.LoadOf(id, cameras) / p.Nodes[id].Capacity
			if loads[id] > loads[hi] {
				hi = id
			}
			if loads[id] < loads[lo] {
				lo = id
			}
		}
		if loads[hi]-loads[lo] < deadBand {
			break
		}
		var candidate int64 = -1
		for c, pl := range p.Placed {
			cam, ok := cameras[c]
			if !ok || pl.Node != hi || !cam.Labels.SubsetOf(p.Nodes[lo].Labels) {
				continue
			}
			if candidate < 0 || cam.Load < cameras[candidate].Load || (cam.Load == cameras[candidate].Load && c < candidate) {
				candidate = c
			}
		}
		if candidate < 0 {
			break
		}
		if p.LoadOf(lo, cameras)+cameras[candidate].Load > p.Nodes[lo].Capacity {
			break
		}
		rev, err := p.rev()
		if err != nil {
			return moves, err
		}
		if err := p.store(candidate, Placement{lo, fmt.Sprintf("rebalance from %s (spread %.0f%%)", hi, (loads[hi]-loads[lo])*100), rev}); err != nil {
			return moves, err
		}
		moves = append(moves, Move{candidate, hi, lo})
	}
	return moves, nil
}

// CheckInvariants is the property test's oracle.
func CheckInvariants(p *Placer, cameras map[int64]PCamera) error {
	for c, pl := range p.Placed {
		n, ok := p.Nodes[pl.Node]
		if !ok {
			return fmt.Errorf("camera %d placed on a Node that does not exist", c)
		}
		if !cameras[c].Labels.SubsetOf(n.Labels) {
			return fmt.Errorf("camera %d violates its constraint", c)
		}
	}
	for _, n := range p.Nodes {
		if p.LoadOf(n.ID, cameras) > n.Capacity+1e-9 {
			return fmt.Errorf("%s over capacity", n.ID)
		}
	}
	return nil
}

File: М11_ClusterVMS/clustervms/cluster/placement.py
"""Lesson 5 — placement onto Nodes: capacity measured, constraints as
labels, the placement STORED in Variables (placement/<camera>), and one
rule with a property test: adding a Node moves nothing.

Rebalance is explicit, budgeted, observable, interruptible, with a dead
band — and each move is the one two-writer operation in the module, so
the caller performs it with the epoch: stop on the old Node, start on the
new, never both at once without a fence.
"""
from __future__ import annotations

from dataclasses import dataclass

from .variables import Variables


@dataclass(frozen=True)
class Camera:
    id: int
    load: float                       # in units of the probe's per-pipeline increment I
    labels: frozenset = frozenset()   # e.g. {"vlan:cctv-b"}


@dataclass
class Node:
    id: str
    capacity: float                   # (budget − B) / I from shard-memory-probe.py
    labels: frozenset = frozenset()


@dataclass
class Placement:
    node: str
    reason: str
    rev: int


class Placer:
    """Placements live in Variables under placement/<camera_id>: small,
    one writer (the placement service), exact lookups only."""

    def __init__(self, nodes: dict[str, Node], vars_: Variables):
        self.nodes = nodes
        self.vars = vars_
        self.placed: dict[int, Placement] = {}
        self._load()

    def _load(self):
        for path in self.vars.list("placement/"):
            items, _ = self.vars.get(path)
            if items:
                self.placed[int(path.split("/")[1])] = Placement(items["node"], items["reason"], int(items["rev"]))

    def _store(self, cam_id: int, pl: Placement):
        _, idx = self.vars.get(f"placement/{cam_id}")
        self.vars.put(f"placement/{cam_id}", {"node": pl.node, "reason": pl.reason, "rev": pl.rev}, cas=idx)
        self.placed[cam_id] = pl

    def _rev(self) -> int:
        items, _ = self.vars.get("placement")
        rev = int(items["rev"]) + 1 if items else 1
        self.vars.put("placement", {"rev": rev})
        return rev

    # -- queries --------------------------------------------------------------
    def load_of(self, node_id: str, cameras: dict[int, Camera]) -> float:
        return sum(cameras[c].load for c, p in self.placed.items() if p.node == node_id and c in cameras)

    def eligible(self, cam: Camera) -> list[Node]:
        return [n for n in self.nodes.values() if cam.labels <= n.labels]

    # -- placement ------------------------------------------------------------
    def place(self, cam: Camera, cameras: dict[int, Camera]) -> Placement | None:
        """Place ONE new camera. Existing placements are never touched."""
        if cam.id in self.placed:
            return self.placed[cam.id]
        best, best_free = None, 0.0
        for n in sorted(self.eligible(cam), key=lambda n: n.id):       # deterministic tiebreak
            free = n.capacity - self.load_of(n.id, cameras)
            if free >= cam.load and free > best_free:
                best, best_free = n, free
        if best is None:
            return None                           # "the system is full" — never "Node 3 is full"
        pl = Placement(best.id, f"most free capacity ({best_free:.1f}) among {len(self.eligible(cam))} eligible",
                       self._rev())
        self._store(cam.id, pl)
        return pl

    def remove(self, cam_id: int) -> None:
        if cam_id in self.placed:
            _, idx = self.vars.get(f"placement/{cam_id}")
            self.vars.put(f"placement/{cam_id}", {"node": "", "reason": "removed", "rev": self._rev()}, cas=idx)
            del self.placed[cam_id]

    def add_node(self, node: Node) -> None:
        self.nodes[node.id] = node                # nothing moves. That is the rule.

    def retire_node(self, node_id: str, cameras: dict[int, Camera]) -> list[int]:
        """A Node retired ON PURPOSE (a failover moves the Node, not its cameras)."""
        del self.nodes[node_id]
        orphans = [c for c, p in self.placed.items() if p.node == node_id]
        homeless = []
        for c in orphans:
            del self.placed[c]
            if self.place(cameras[c], cameras) is None:
                homeless.append(c)
        return homeless

    # -- rebalance ------------------------------------------------------------
    def rebalance(self, cameras: dict[int, Camera], budget: int, dead_band: float = 0.10) -> list[tuple[int, str, str]]:
        moves = []
        for _ in range(budget):
            loads = {n: self.load_of(n, cameras) / self.nodes[n].capacity for n in self.nodes}
            if not loads:
                break
            hi = max(loads, key=loads.get); lo = min(loads, key=loads.get)
            if loads[hi] - loads[lo] < dead_band:
                break
            candidates = [c for c, p in self.placed.items() if p.node == hi and c in cameras
                          and cameras[c].labels <= self.nodes[lo].labels]
            if not candidates:
                break
            c = min(candidates, key=lambda c: cameras[c].load)
            if self.load_of(lo, cameras) + cameras[c].load > self.nodes[lo].capacity:
                break
            self._store(c, Placement(lo, f"rebalance from {hi} (spread {loads[hi]-loads[lo]:.0%})", self._rev()))
            moves.append((c, hi, lo))
        return moves


def check_invariants(p: Placer, cameras: dict[int, Camera]) -> None:
    for cid, pl in p.placed.items():
        assert pl.node in p.nodes, f"camera {cid} placed on a Node that does not exist"
        assert cameras[cid].labels <= p.nodes[pl.node].labels, f"camera {cid} violates its constraint"
    for n in p.nodes.values():
        assert p.load_of(n.id, cameras) <= n.capacity + 1e-9, f"{n.id} over capacity"
